mne/io/bti/bti.py

```diff
@@ -913,6 +913,10 @@ def _read_bti_header(pdf_fname, config_fname, sort_by_ch_name=True):
         chans_cfg = [c for c in cfg['chs'] if c['chan_no']
                      in [c_['chan_no'] for c_ in chans]]
 
+        # sort chans_cfg and chans
+        chans = sorted(chans, key=lambda k: k['chan_no'])
+        chans_cfg = sorted(chans_cfg, key=lambda k: k['chan_no'])
```

Comment:
If the `chan_no` always corresponds to the index in the data, then this is the right thing to do.

But it's possible a different scheme is used, where `chans_cfg` or `chans` is already ordered to match the data, and `chan_no` does not necessarily give the index to the data (e.g., `chans[0]` corresponds to the first "row" of stored data, but `chans[0]['chan_no']` could be some other number like 2), then it's not. Any ideas from the format?

@dengemann I think you originally wrote the BTi code, any idea about this, or how to check?

Reply:
Since the subject I ran across this error on was the first to throw an exception, I suspect (but have not confirmed) that the order of 'chan_no' does much the row indices in other subjects. In this one, you are correct that the initial and 'chan_no' do not correlate. For example, in my data, 'chans' is derived as done by MNE while 'nchans' is the sorting result:

>>> chans[0]
{'chan_label': 'A22', 'chan_no': 8, 'attributes': 0, 'scale': 1.0, 'yaxis_label': 'T', 'valid_min_max': 1, 'ymin': -4.868604127669051e-12, 'ymax': 4.3606116638294954e-12, 'index': 0, 'checksum': -2876, 'off_flag': 'OFF+', 'offset': 0.0}

>>> nchans[0]
{'chan_label': 'TRIGGER', 'chan_no': 1, 'attributes': 2, 'scale': 1.0, 'yaxis_label': 'bit', 'valid_min_max': 1, 'ymin': 0.0, 'ymax': 8192.0, 'index': 303, 'checksum': -1615, 'off_flag': 'OFF-', 'offset': 0.0}

In the original 'chans', the TRIGGER (chan_no=1) channel is at index 303:
>>> chans[303]
{'chan_label': 'TRIGGER', 'chan_no': 1, 'attributes': 2, 'scale': 1.0, 'yaxis_label': 'bit', 'valid_min_max': 1, 'ymin': 0.0, 'ymax': 8192.0, 'index': 303, 'checksum': -1615, 'off_flag': 'OFF-', 'offset': 0.0}

The sorting maintains the integrity of full dictionary, so as long as everything is referenced by 'chan_no' and not by the index, it would be fine. If however the index is used without comparing chan_no, it could introduce a mismatch...

Reply:
Sorting the chans and chans_cfg using chan_no should be fine. I used chans.sort(key=lambda x: x['chan_no'])
and chans_cfg.sort(key=lambda x: x['chan_no']), which is probably the same as sorted().
For the background: the config-file contains all channels showing up in a run, a data file (pdf) usually contains just a subset of channels. The value for a channel (chan_no) in a time slice is given by its index (stored in the pdf)
and the (chan_no, chan_index) relation must be maintained, but the sequence of the channel structures
(dftk_channel in the config and dftk_channel_ref in the pdf) is arbitrary. It just happens that during acquisition
channels are stored in the same sequence as in the config file.

Reply:
Great, thanks for the background @ebeich and the +1 (as I understand it) for this change